collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormsDetailsFetcher.kt

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2018 Nafundi
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.odk.collect.android.formmanagement
+
+import org.odk.collect.analytics.Analytics.Companion.log
+import org.odk.collect.android.analytics.AnalyticsEvents
+import org.odk.collect.android.openrosa.OpenRosaFormSource
+import org.odk.collect.android.utilities.FormUtils
+import org.odk.collect.android.utilities.WebCredentialsUtils
+import org.odk.collect.forms.Form
+import org.odk.collect.forms.FormSource
+import org.odk.collect.forms.FormSourceException
+import org.odk.collect.forms.FormsRepository
+import org.odk.collect.forms.ManifestFile
+import org.odk.collect.forms.MediaFile
+import org.odk.collect.shared.strings.Md5.getMd5Hash
+import timber.log.Timber
+import java.io.File
+
+/**
+ * Open to allow mocking (used in existing Java tests)
+ */
+open class ServerFormsDetailsFetcher(
+    private val formsRepository: FormsRepository,
+    private val formSource: FormSource,
+    private val diskFormsSynchronizer: DiskFormsSynchronizer
+) {
+    open fun updateUrl(url: String) {
+        (formSource as OpenRosaFormSource).updateUrl(url)
+    }
+
+    open fun updateCredentials(webCredentialsUtils: WebCredentialsUtils) {
+        (formSource as OpenRosaFormSource).updateWebCredentialsUtils(webCredentialsUtils)
+    }
+
+    @Throws(FormSourceException::class)
+    open fun fetchFormDetails(): List<ServerFormDetails> {
+        diskFormsSynchronizer.synchronize()
+
+        val formList = formSource.fetchFormList()
+        return formList.map { listItem ->
+            val manifestFile = listItem.manifestURL.let {
+                if (it != null) {
+                    getManifestFile(formSource, it)
+                } else {
+                    null
+                }
+            }
+
+            val forms = formsRepository.getAllNotDeletedByFormId(listItem.formID)
+            val thisFormAlreadyDownloaded = forms.isNotEmpty()
+            val isNewerFormVersionAvailable = listItem.hashWithPrefix.let {
+                if (it == null || !it.startsWith("md5:")) {
```

Comment:
I would move checking for that prefix `md5:` to [OpenRosaResponseParserImpl](https://github.com/getodk/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/openrosa/OpenRosaResponseParserImpl.java#L106) so that we deal with it only there. If that prefix does not exist we should return null like in case of empty value. What do you think?

Reply:
That's an interesting idea. I'll have a play around with it.

Reply:
I would even go the extra mile and remove that prefix there so that in `FormListItem` we have it without `md5:` and we don't need to call `substring` every time we use the hash.

Reply:
...or pass it to `FormListItem` with the prefix but have a custom getter which would ignore it.

Reply:
Yeah I was also thinking about a new type (like a `Hash`) that enforces the prefix but means we don't have to deal with it everywhere. I'll play around with a few of these ideas, see what I like, and then we can review!

Reply:
If you decide to implement changes in `OpenRosaResponseParserImpl` can convert it to Kotlin as well,

Reply:
Ok done!